From a8237d70f7beb937f0dc457544dd084e20c53056 Mon Sep 17 00:00:00 2001
From: Bakkeby <bakkeby@gmail.com>
Date: Mon, 1 Jul 2024 23:06:53 +0200
Subject: [PATCH 2/2] Adding cursorwarp patch on top of banish

---
 dwm.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/dwm.c b/dwm.c
index 7fcab53..c094164 100644
--- a/dwm.c
+++ b/dwm.c
@@ -236,6 +236,7 @@ static void updatetitle(Client *c);
 static void updatewindowtype(Client *c);
 static void updatewmhints(Client *c);
 static void view(const Arg *arg);
+static void warp(Client *c);
 static Client *wintoclient(Window w);
 static Monitor *wintomon(Window w);
 static int xerror(Display *dpy, XErrorEvent *ee);
@@ -865,6 +866,8 @@ focusmon(const Arg *arg)
 	unfocus(selmon->sel, 0);
 	selmon = m;
 	focus(NULL);
+	if (m->sel)
+		warp(m->sel)
 }
 
 void
@@ -890,6 +893,7 @@ focusstack(const Arg *arg)
 	if (c) {
 		focus(c);
 		restack(selmon);
+		warp(c)
 	}
 }
 
@@ -1167,6 +1171,8 @@ manage(Window w, XWindowAttributes *wa)
 	c->mon->sel = c;
 	arrange(c->mon);
 	XMapWindow(dpy, c->win);
+	if (c && c->mon == selmon && ISVISIBLE(c))
+		warp(c);
 	focus(NULL);
 }
 
@@ -1936,6 +1942,8 @@ unmanage(Client *c, int destroyed)
 	focus(NULL);
 	updateclientlist();
 	arrange(m);
+	if (m->sel && m == selmon)
+		warp(m->sel);
 }
 
 void
@@ -2199,6 +2207,18 @@ view(const Arg *arg)
 	arrange(selmon);
 }
 
+void
+warp(Client *c)
+{
+	if (cursor_hidden) {
+		mouse_x = c->x + c->w/2;
+		mouse_y = c->y + c->h/2;
+		return;
+	}
+
+	XWarpPointer(dpy, None, c->win, 0, 0, 0, 0, c->w/2, c->h/2);
+}
+
 Client *
 wintoclient(Window w)
 {
-- 
2.50.1

