From 31bfeba5a8993201bddf8736f0e236c3d5273384 Mon Sep 17 00:00:00 2001
From: Bakkeby <bakkeby@gmail.com>
Date: Wed, 26 Jun 2024 09:46:49 +0200
Subject: [PATCH] Adding desktop patch

---
 dwm.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/dwm.c b/dwm.c
index 4465af1..328540d 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1737,12 +1737,14 @@ toggletag(const Arg *arg)
 void
 toggleview(const Arg *arg)
 {
+	Monitor *m;
 	unsigned int newtagset = selmon->tagset[selmon->seltags] ^ (arg->ui & TAGMASK);
 
 	if (newtagset) {
-		selmon->tagset[selmon->seltags] = newtagset;
+		for (m = mons; m; m = m->next)
+			m->tagset[m->seltags] = newtagset;
 		focus(NULL);
-		arrange(selmon);
+		arrange(NULL);
 	}
 }
 
@@ -2035,13 +2037,16 @@ updatewmhints(Client *c)
 void
 view(const Arg *arg)
 {
+	Monitor *m;
 	if ((arg->ui & TAGMASK) == selmon->tagset[selmon->seltags])
 		return;
-	selmon->seltags ^= 1; /* toggle sel tagset */
+	for (m = mons; m; m = m->next)
+		m->seltags ^= 1; /* toggle sel tagset */
 	if (arg->ui & TAGMASK)
-		selmon->tagset[selmon->seltags] = arg->ui & TAGMASK;
+		for (m = mons; m; m = m->next)
+			m->tagset[m->seltags] = arg->ui & TAGMASK;
 	focus(NULL);
-	arrange(selmon);
+	arrange(NULL);
 }
 
 Client *
-- 
2.45.2

