From 026551d381929c1097dd1b8a1756def7c5dcadbe Mon Sep 17 00:00:00 2001
From: bakkeby <bakkeby@gmail.com>
Date: Tue, 16 Feb 2021 10:22:40 +0100
Subject: [PATCH 2/2] f

---
 dwm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dwm.c b/dwm.c
index 8b25561..74dc566 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1526,7 +1526,7 @@ rioresize(const Arg *arg)
 		riodraw(c, slopresizestyle);
 }
 
-/* spawn a new window and drag out an area using slop to postiion it */
+/* spawn a new window and drag out an area using slop to position it */
 void
 riospawn(const Arg *arg)
 {
-- 
2.19.1

